README.md

<!-- README.md is generated from README.Rmd. Please edit that file -->

# glmmboot

<!-- badges: start -->

[![Travis build
status](https://travis-ci.org/colmanhumphrey/glmmboot.svg?branch=master)](https://travis-ci.org/colmanhumphrey/glmmboot)
[![Codecov test
coverage](https://codecov.io/gh/colmanhumphrey/glmmboot/branch/master/graph/badge.svg)](https://codecov.io/gh/colmanhumphrey/glmmboot?branch=master)
[![CRAN
status](https://www.r-pkg.org/badges/version/glmmboot)](https://cran.r-project.org/package=glmmboot)
<!-- badges: end -->

## Overview

glmmboot provides a simple interface for creating non-parametric
bootstrap confidence intervals using a wide set of models. The primary
function is `bootstrap_model`, which has three primary arguments:

-   `base_model`: the model run on the full data as you normally would,
    prior to bootstrapping
-   `base_data`: the dataset used
-   `resamples`: how many bootstrap resamples you wish to perform

Another function, `bootstrap_ci`, converts output from bootstrap model
runs into confidence intervals and p-values. By default,
`bootstrap_model` calls `bootstrap_ci`.

## Types of bootstrapping

For models with random effects:

-   the default (and recommended) behaviour will be to block sample over
    the effect with the largest entropy (generally the one with the most
    levels)
-   it’s also possible to specify multiple random effects to block
    sample over

With no random effects, performs case resampling: resamples each row
with replacement.

All of these are considered non-parametric.

## Requirements:

1.  the model should work with the function `update`, to change the data
2.  the coefficients are extractable using `coef(summary(model))`

-   either directly, i.e. this gives a matrix
-   or it’s a list of matrices; this includes e.g. zero-inflated models,
    which produce two matrices of coefficients

## Parallel

It may be desired to run this package in parallel. The best way is to
use the `future` backend, which uses `future.apply::future_lapply`. You
do that by specifying the backend through the `future::plan` setup, and
then setting `parallelism = "future"`. It’s quite possible you’ll want
to pass the package used to build the model to the argument
`future_packages`. See the Quick Use vignette for more.

It’s also easy to use `parallel::mclapply`; again, see the Quick Use
vignette.

## Installation

glmmboot is on CRAN, so you can install it normally:

``` r
install.packages("glmmboot")
```

Or the development version:

``` r
## install.packages("devtools")
devtools::install_github("ColmanHumphrey/glmmboot")
```

## Example: glm (no random effect)

We’ll provide a quick example using glm. First we’ll set up some data:

``` r
set.seed(15278086)
x1 <- rnorm(50)
x2 <- runif(50)

expit <- function(x){exp(x) / (1 + exp(x))}

y_mean <- expit(0.2 - 0.3 * x1 + 0.4 * x2)

y <- rbinom(50, 1, prob = y_mean)

sample_frame <- data.frame(x1 = x1, x2 = x2, y = y)
```

Typically this model is fit with logistic regression:

``` r
base_run <- glm(y ~ x1 + x2,
                family = binomial(link = 'logit'),
                data = sample_frame)
summary(base_run)
# 
# Call:
# glm(formula = y ~ x1 + x2, family = binomial(link = "logit"), 
#     data = sample_frame)
# 
# Deviance Residuals: 
#     Min       1Q   Median       3Q      Max  
# -1.6819  -1.2340   0.7048   0.9389   1.3213  
# 
# Coefficients:
#             Estimate Std. Error z value Pr(>|z|)
# (Intercept)  -0.1161     0.5890  -0.197    0.844
# x1           -0.5147     0.3387  -1.519    0.129
# x2            1.0933     1.0065   1.086    0.277
# 
# (Dispersion parameter for binomial family taken to be 1)
# 
#     Null deviance: 65.342  on 49  degrees of freedom
# Residual deviance: 61.944  on 47  degrees of freedom
# AIC: 67.944
# 
# Number of Fisher Scoring iterations: 4
```

Let’s run a bootstrap.

``` r
library(glmmboot)
boot_results <- bootstrap_model(base_model = base_run,
                                base_data = sample_frame,
                                resamples = 999)
```

And the results:

``` r
print(boot_results)
#               estimate boot 2.5% boot 97.5% boot p_value base p_value base 2.5%
# (Intercept) -0.1160896   -1.2295     0.9809        0.830       0.8446   -1.3010
# x1          -0.5146778   -1.1245     0.0455        0.076       0.1353   -1.1961
# x2           1.0932707   -0.7517     3.1328        0.284       0.2829   -0.9315
#             base 97.5% boot/base width
# (Intercept)     1.0688       0.9327523
# x1              0.1667       0.8584962
# x2              3.1181       0.9592352
```

The estimates are the same, since we just pull from the base model. The
intervals are similar to the base model, although slightly narrower:
typical logistic regression is fractionally conservative at `N = 50`.

An example with a zero-inflated model (from the `glmmTMB` docs):

``` r
## we'll skip this if glmmTMB not available
library(glmmTMB)

owls <- transform(Owls,
                  nest = reorder(Nest, NegPerChick),
                  ncalls = SiblingNegotiation,
                  ft = FoodTreatment)

fit_zipoisson <- glmmTMB(
    ncalls ~ (ft + ArrivalTime) * SexParent +
        offset(log(BroodSize)) + (1 | nest),
    data = owls,
    ziformula = ~1,
    family = poisson)

summary(fit_zipoisson)
#  Family: poisson  ( log )
# Formula:          
# ncalls ~ (ft + ArrivalTime) * SexParent + offset(log(BroodSize)) +  
#     (1 | nest)
# Zero inflation:          ~1
# Data: owls
# 
#      AIC      BIC   logLik deviance df.resid 
#   4015.6   4050.8  -1999.8   3999.6      591 
# 
# Random effects:
# 
# Conditional model:
#  Groups Name        Variance Std.Dev.
#  nest   (Intercept) 0.1294   0.3597  
# Number of obs: 599, groups:  nest, 27
# 
# Conditional model:
#                           Estimate Std. Error z value Pr(>|z|)    
# (Intercept)                2.53995    0.35656   7.123 1.05e-12 ***
# ftSatiated                -0.29111    0.05961  -4.884 1.04e-06 ***
# ArrivalTime               -0.06808    0.01427  -4.771 1.84e-06 ***
# SexParentMale              0.44885    0.45002   0.997    0.319    
# ftSatiated:SexParentMale   0.10473    0.07286   1.437    0.151    
# ArrivalTime:SexParentMale -0.02140    0.01835  -1.166    0.244    
# ---
# Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
# 
# Zero-inflation model:
#             Estimate Std. Error z value Pr(>|z|)    
# (Intercept) -1.05753    0.09412  -11.24   <2e-16 ***
# ---
# Signif. codes:  0 '***' 0.001 '**' 0.01 '*' 0.05 '.' 0.1 ' ' 1
```

Let’s run the bootstrap (ignore the actual results, 3 resamples is
basically meaningless - just for illustration):

``` r
zi_results <- bootstrap_model(base_model = fit_zipoisson,
                              base_data = owls,
                              resamples = 3)

print(zi_results)
# $cond
#                              estimate boot 2.5% boot 97.5% boot p_value
# (Intercept)                2.53994692    1.9197     2.9229          0.5
# ftSatiated                -0.29110639   -0.3058    -0.1889          0.5
# ArrivalTime               -0.06807809   -0.0866    -0.0392          0.5
# SexParentMale              0.44884508    0.1134     1.2690          0.5
# ftSatiated:SexParentMale   0.10472505   -0.1153     0.2804          1.0
# ArrivalTime:SexParentMale -0.02139750   -0.0527    -0.0087          0.5
#                           base p_value base 2.5% base 97.5% boot/base width
# (Intercept)                     0.0000    1.8411     3.2388       0.7177368
# ftSatiated                      0.0000   -0.4079    -0.1743       0.5002454
# ArrivalTime                     0.0000   -0.0960    -0.0401       0.8479791
# SexParentMale                   0.3186   -0.4332     1.3309       0.6550388
# ftSatiated:SexParentMale        0.1506   -0.0381     0.2475       1.3852712
# ArrivalTime:SexParentMale       0.2436   -0.0574     0.0146       0.6116518
# 
# $zi
#              estimate boot 2.5% boot 97.5% boot p_value base p_value base 2.5%
# (Intercept) -1.057534   -1.0575      -0.84          0.5            0    -1.242
#             base 97.5% boot/base width
# (Intercept)    -0.8731       0.5895082
```

We could also have run this with the `future.apply` backend:

``` r
library(future.apply)
plan("multiprocess")

zi_results <- bootstrap_model(base_model = fit_zipoisson,
                              base_data = owls,
                              resamples = 1000,
                              parallelism = "future",
                              future_packages = "glmmTMB")
```


NEWS.md
# glmmboot 0.6.0

* Removing previously deprecated functions (deprecated as of 0.4.0)
* Adding conditional evaluation for tests and the vignette (and the readme)
  for `glmmTMB`, in case it's not present / installable on a given system (as
  it's only a suggested package)
* Removing `call. = FALSE` for errors and warnings
* adding `future.seed = TRUE` to `future.apply::future_lapply` call for better RNG
* Minor vignette and README copy edits

# glmmboot 0.5.1

* Changing a `donttest` to a `dontrun` within an example.

# glmmboot 0.5.0

* if `parallelism = "parallel"` but `num_cores` is left as NULL, we default to using `parallel::detectCores() - 1L` cores. This was in fact the documented behaviour previously, just sadly not the actual behaviour
* `bootstrap_model` (and internal function `bootstrap_runner`) accepts an argument `future_packages` that will be passed along to `future.apply::future.lapply`; this is needed for futures that don't share memory, because the required global isn't visible (when using S3 generics...). This is also passed along to README/vignettes/tests etc.
* Some documentation edits, shouldn't affect functionality
* sampling message shown by default in interactive settings
* fixing all incorrect uses of `class` (new matrix class is the biggest change) - mostly was in tests!

# glmmboot 0.4.0

API changes:

* `BootGlmm` is now `bootstrap_model` as the primary exported function
* `BootCI` is now `bootstrap_ci`
* Both of these functions should now work smoothly if your model returns a list of matrix coefficients; it will perform the bootstrap sampling on all of them
* You basically have to supply `base_data` to `bootstrap_model`; to reflect this, it's now the second argument, prior to `resamples`
* parallel functionality is improve, now supporting the `future.apply::future_lapply` backend (and some slightly more robust versions of e.g. `parallel::mclapply` if desired)
* Added code coverage

# glmmboot 0.3.0

* For general narrowness bias, now resampling k-1 levels of random effects / n-1 rows of regular data (depending on existence of random effects). This can be turned off with narrowness_avoid = FALSE in BootGlmm.

# glmmboot 0.2.0

* Added a warning if data not supplied explicitly
* Changed two-sided p-value calculation to more tightly respect the fact that the bootstrap t-values
  intentionally are not forced to be symmetric.

# Travis CI:
[![Travis-CI Build Status](https://travis-ci.org/ColmanHumphrey/glmmboot.svg?branch=master)](https://travis-ci.org/ColmanHumphrey/glmmboot)

# glmmboot 0.1.2

* Added a `NEWS.md` file to track changes to the package.
* Initial release of glmmboot


tests/testthat/test_BootCI.R
context('testing BootCI and related tools')

test_that('BootCI returns matrix when given list from BootGlmm', {
    x <- rnorm(10)
    y <- rnorm(10)    
    simple_model <- lm(y ~ x)

    list_out = BootGlmm(base_model = simple_model, resamples = 20,
                        return_coefs_instead = TRUE)

    expect_equal(class(BootCI(list_out$base_coef_se,
                              list_out$resampled_coef_se)),
                 'matrix')
})

test_that('CombineResampledLists gives same result for list of lists, or one by one', {
    x <- rnorm(10)
    y <- rnorm(10)    
    simple_model <- lm(y ~ x)

    list_out1 = BootGlmm(base_model = simple_model, resamples = 20,
                        return_coefs_instead = TRUE)
    list_out2 = BootGlmm(base_model = simple_model, resamples = 20,
                        return_coefs_instead = TRUE)
    list_out3 = BootGlmm(base_model = simple_model, resamples = 20,
                         return_coefs_instead = TRUE)

    expect_equal(CombineResampledLists(list_out1, list_out2, list_out3),
                 CombineResampledLists(list(list_out1, list_out2, list_out3)))
})
    


tests/testthat/test_GetRand.R
context('Testing that GetRand behaves as expected')

test_that('GetRand gives correct output', {

    expect_equal(GetRand('y ~ age + (1 | school)'),
                 'school')

    expect_equal(GetRand('y ~ income + (1 | school) + (1 | school:section)'),
                 c('school', 'section'))
    
    expect_equal(GetRand('y ~ income + (1 | school) + (1 | school/section)'),
                 c('school', 'section'))
    
    expect_equal(GetRand(as.formula('y ~ x + (1 | z)')),
                 'z')

    ## i.e. empty vector
    expect_equal(GetRand('y ~ x'),
                 vector(mode = 'character'))
}


tests/testthat/test_BootGlmm.R
context('Testing BootGlmm (and DetectCores, and CalcEntropy)')

test_that('BootGlmm returns matrix when asked, lists when asked', {
    x <- rnorm(10)
    y <- rnorm(10)    
    simple_model <- lm(y ~ x)
    
    expect_equal(
        class(BootGlmm(base_model = simple_model, resamples = 20)),
        'matrix')
    expect_equal(
        class(BootGlmm(base_model = simple_model, resamples = 20,
                       return_coefs_instead = TRUE)),
        'list')
})

test_that('BootGlmm fails when model has no data and none is supplied.', {
    x <- rnorm(10)
    y <- rnorm(10)    
    simple_model <- lm(y ~ x)

    simple_model_nodata <- simple_model
    simple_model_nodata$model <- NULL
    
    expect_error(
        BootGlmm(base_model = simple_model_nodata, resamples = 20)
    )
})


    


R/data.R
#' Simulated data containing three fixed effects and one random effect
#'
#' A small normal dataset with a proportional
#' outcome to illustrate the use of this package.
#' The outcome has mean
#' \code{expit(0.5 + 0.1 * x_var1 + 0.2 * x_var2 + 0.3 * x_var3 + SUBJ_VAL)}
#' where SUBJ_VAL are the values of the random effect
#' The SD of y is then shrunk by 0.9 relative to a binomial distribution,
#' and then beta values are generated. Arbitrarily close to the endpoints
#' gives zeros and ones.
#'
#' @format A data frame with 300 rows and 4 variables:
#' \describe{
#'   \item{x_var1}{independent normally distributed variable}
#'   \item{x_var2}{independent normally distributed variable}
#'   \item{x_var3}{independent normally distributed variable}
#'   \item{subj}{levels of random effect}
#'   \item{y}{outcome: lives in interval [0,1]}
#' }
#'
"test_data"


R/BootCI.R
#' Enter first level estimates and second level estimates,
#' get bootstrap interval, from the pivotal bootstrap t
#' (Efron and Tibshirani 1994, also endorsed
#' by Hesterberg 2015).
#'
#' @param base_coef_se
#'   Estimates and SEs from full sample. In matrix form:
#'   i.e. a (p+1) x 2 matrix, first column is estimates,
#'   second is standard errors. This
#'   is the output from using:
#'   coef(summary(model_output))[,1:2, drop = FALSE]
#'   or
#'   coef(summary(model_output))$cond[,1:2, drop = FALSE]
#'   if model_output is the output from a random
#'   effects model (some may not have cond as the correct pull).
#'
#' @param resampled_coef_se
#'   List of estimates and SEs from the bootstrapped resamples,
#'   each list entry has the same format as the base_coef_se above.
#'
#' @param orig_df
#'   Degrees of freedom to use to calculate the
#'   t-values used for the base interval.
#'
#' @param alp_level
#'   level of CI - if you fill in \code{probs}, will use those instead
#'
#' @param probs
#'   Default NULL, and will use alp_level to set
#'   endpoints. Else will calculate these CI endpoints.
#'
#' @export
#'
#' @return A matrix containing:
#'   Estimates;
#'   Bootstrap interval endpoints;
#'   Bootstrap p-value;
#'   Base p-value;
#'   Base interval endpoints;
#'   Relative width of bootstrap interval to base
#'
#' @examples
#' x <- rnorm(20)
#' y <- rnorm(20) + x
#' xy_data = data.frame(x = x, y = y)
#' first_model <- lm(y ~ x, data = xy_data)
#' out_list <- BootGlmm(first_model, 20, base_data = xy_data, return_coefs_instead = TRUE)
#' BootCI(out_list$base_coef_se,
#'        out_list$resampled_coef_se)
#'
#' \donttest{
#'   data(test_data)
#'   library(glmmTMB)
#'   ## where subj is some random effect
#'   test_model <- glmmTMB(y ~ x_var1 + (1 | subj), data = test_data, family = binomial)
#'   output_lists <- BootGlmm(test_model, 199, base_data = test_data, return_coefs_instead = TRUE)
#'   BootCI(output_lists$base_coef_se,
#'          output_lists$resampled_coef_se)
#' }
BootCI <- function(base_coef_se = NULL,
                   resampled_coef_se = NULL,
                   orig_df = NULL,
                   alp_level = 0.05,
                   probs = NULL){
    if(is.null(probs)){
        if(alp_level < 0 | alp_level > 0.5){
            stop("Can't calculate a two-sided CI with this alpha value, must be in (0, 0.5)")
        }
        probs = sort(c(alp_level / 2, 1 - alp_level/2))
        ## probs = c(alp_level / 2, 1 - alp_level/2)
    }
    if(max(probs) > 1 | min(probs) < 0){
        stop("Probabilities should be in (0,1)")
    }

    base_row_names = rownames(base_coef_se)
    all_row_names <- lapply(resampled_coef_se, rownames)

    name_match <- unlist(lapply(all_row_names, function(x){mean(x == base_row_names) == 1}))

    if(mean(name_match) != 1){
        stop("Naming mismatch from base to list of coefs")
    }

    conf_ind <- matrix(NA, nrow = length(base_row_names),
                       ncol = length(probs) + 1)
    rownames(conf_ind) = base_row_names

    ## actual work here...:
    for(var in base_row_names){
        base_est = base_coef_se[var,1]
        base_se = base_coef_se[var,2]

        rep_ests = unlist(lapply(resampled_coef_se, function(x){x[var,1]}))
        rep_ses = unlist(lapply(resampled_coef_se, function(x){x[var,2]}))

        ## thanks Tim!
        t_boot <- quantile( (rep_ests - base_est) / rep_ses, 1 - probs, type = 6)
        ## can't have the CI not containing the estimate...!
        ## shouldn't happen too often...
        t_boot[1] = ifelse(t_boot[1] < 0, 0, t_boot[1])
        t_boot[2] = ifelse(t_boot[2] > 0, 0, t_boot[2])

        result <- base_est - t_boot * base_se

        ## p val:
        p_t = base_est / base_se

        p_val_num_left = sum((rep_ests - base_est) / rep_ses <=  p_t) + 1
        p_val_num_right = sum((rep_ests - base_est) / rep_ses >=  p_t) + 1

        p_val_num = 2 * min(p_val_num_left, p_val_num_right)

        p_val = p_val_num / (length(rep_ests) + 1)

        conf_ind[var,] = c(result, p_val)
    }
    colnames(conf_ind) = c(paste0('boot ',names(quantile(1, probs))), ' boot p_value')

    use_df = ifelse(is.null(orig_df), Inf, orig_df)
    t_base_vals = qt(probs, df = use_df)
    base_mat <- cbind(sapply(t_base_vals, function(x){base_coef_se[,1] + x * base_coef_se[,2]}),
                      2 * pt(-abs(base_coef_se[,1] / base_coef_se[,2]),
                                              df = use_df))
    colnames(base_mat) = c(paste0('base ',names(quantile(1, probs))), 'base p_value')

    ## p_star <- GenStar(conf_ind[,3])

    ret_matrix <- cbind('estimate' = base_coef_se[,1], round(conf_ind, 4),
                        ## p_star,
                        round(base_mat[,c(3,1,2)], 4),
                        'boot/base width' = (conf_ind[,2] - conf_ind[,1]) / (base_mat[,2] - base_mat[,1]))

    return(ret_matrix)
}

#' When running BootGlmm in distributed fashion,
#' combines output
#'
#' If you run BootGlmm on e.g. a grid of computers,
#' set return_coefs_instead = TRUE for each.
#' Then enter them all here. Either just list them out,
#' or put them into one list and enter them.
#'
#' @param ...
#'   Say our output from BootGlmm from three separate computers is
#'   output_list1,
#'   output_list2,
#'   output_list3
#'   We can run: CombineResampledLists(output_list1, output_list2, output_list3)
#'   OR: create a list of lists:
#'   output_list_list <- list(output_list1, output_list2, output_list3)
#'   and then: CombineResampledLists(output_list_list)
#'
#' @param return_combined_list
#'   Logical, default FALSE. TRUE if you want the combined
#'   list of lists, FALSE for just the output from BootCI applied to it.
#'
#' @export
#'
#' @return Returns the same output as BootCI by default,
#'   or the combined list (as if you had just run BootGlmm once with
#'   all resamples) if return_combined_list = TRUE
#'
#' @examples
#' \donttest{
#'   data(test_data)
#'   library(glmmTMB)
#'   ## where subj is some RE
#'   test_model <- glmmTMB(y ~ x_var1 + (1 | subj), data = test_data, family = binomial)
#'   output_list1 <- BootGlmm(test_model, 99, base_data = test_data, return_coefs_instead = TRUE)
#'   output_list2 <- BootGlmm(test_model, 100, base_data = test_data, return_coefs_instead = TRUE)
#'   output_list3 <- BootGlmm(test_model, 100, base_data = test_data, return_coefs_instead = TRUE)
#'   CombineResampledLists(output_list1, output_list2, output_list3)
#'
#'   num_blocks = 10
#'   num_total_resamples = 299
#'   reg_list <- list()
#'   for(i in 1:num_blocks){
#'       if(i < num_blocks){
#'           block_resamples = floor((num_total_resamples + 1)/num_blocks)
#'       } else {
#'           block_resamples = floor((num_total_resamples + 1)/num_blocks - 1)
#'       }
#'       reg_list[[i]] = BootGlmm(test_model,
#'                                resamples = block_resamples,
#'                                base_data = test_data,
#'                                return_coefs_instead = TRUE,
#'                                num_cores = 1, ## increase for parallel
#'                                suppress_loading_bar = TRUE)
#'   }
#'   boot_ci1 <- CombineResampledLists(reg_list)
#'   full_list <- CombineResampledLists(reg_list, return_combined_list = TRUE)
#'   boot_ci2 <- BootCI(full_list$base_coef_se, full_list$resampled_coef_se)
#'   identical(boot_ci1, boot_ci2)
#' }
CombineResampledLists <- function(...,
                                  return_combined_list = FALSE){
    input_lists <- list(...)

    if(length(input_lists) == 1){
        input_lists <- input_lists[[1]]
    }

    reg_base_coef <- input_lists[[1]]$base_coef_se

    ## now a list of lists of the resampled:
    listlist_resampled = lapply(input_lists, '[[', 2)
    ## combine to one list
    reg_resampled = do.call(c, listlist_resampled)

    if(return_combined_list){
        return(list(base_coef_se = reg_base_coef,
                    resampled_coef_se = reg_resampled))
    } else {
        block_reg_results <- BootCI(base_coef_se = reg_base_coef,
                                    resampled_coef_se = reg_resampled)

        return(block_reg_results)
    }
}


R/bootstrap_ci.R
#' Generating bootstrap confidence intervals.
#'
#' Enter first level estimates and second level estimates,
#' get bootstrap interval, from the pivotal bootstrap t
#' (Efron and Tibshirani 1994, also endorsed
#' by Hesterberg 2015).
#' @param base_coef_se Estimates and SEs from full sample. In matrix form,
#'   i.e. a \eqn{(p+1) x 2} matrix, first column is estimates,
#'   second is standard errors. This is the output from using:
#'   \code{coef(summary(model_output))[,1:2, drop = FALSE]}
#'   or
#'   \code{coef(summary(model_output))$cond[,1:2, drop = FALSE]}
#'   if \code{model_output} is the output from a random
#'   effects model (some may not have \code{cond} as the correct pull).
#' @param resampled_coef_se List of estimates and SEs from the bootstrapped
#'   resamples, each list entry has the same format as the base_coef_se above.
#' @param orig_df Degrees of freedom to use to calculate the
#'   t-values used for the base interval.
#' @param alpha_level
#'   level of CI - if you fill in \code{probs}, will use those instead
#' @param probs Default \code{NULL}, and will use \code{alpha_level} to set
#'   endpoints. Else will calculate these CI endpoints.
#' @return A matrix containing:
#'   \itemize{
#'    \item Estimates
#'    \item Bootstrap interval endpoints
#'    \item Bootstrap p-value
#'    \item Base p-value
#'    \item Base interval endpoints
#'    \item Relative width of bootstrap interval to base
#'   }
#' @examples
#' x <- rnorm(20)
#' y <- rnorm(20) + x
#' xy_data = data.frame(x = x, y = y)
#' first_model <- lm(y ~ x, data = xy_data)
#' out_list <- bootstrap_model(first_model, base_data = xy_data, 20,
#'                            return_coefs_instead = TRUE)
#' bootstrap_ci(out_list$base_coef_se,
#'              out_list$resampled_coef_se)
#'
#' \donttest{
#'   data(test_data)
#'   library(glmmTMB)
#'   ## where subj is a random effect
#'   test_model <- glmmTMB(y ~ x_var1 + (1 | subj),
#'                         data = test_data, family = binomial)
#'   output_lists <- bootstrap_model(test_model, base_data = test_data, 199,
#'                                   return_coefs_instead = TRUE)
#'   bootstrap_ci(output_lists$base_coef_se,
#'                output_lists$resampled_coef_se)
#' }
#'
#' @export
bootstrap_ci <- function(base_coef_se = NULL,
                         resampled_coef_se = NULL,
                         orig_df = NULL,
                         alpha_level = 0.05,
                         probs = NULL){
    resampled_coef_lists <- lapply(
        1:length(resampled_coef_se[[1]]), function(j){
            lapply(resampled_coef_se, `[[`, j)
        })

    ci_results <- Map(function(base_matrix, resampled_coef_list){
        bootstrap_individual_ci(base_matrix,
                                resampled_coef_list,
                                orig_df = orig_df,
                                alpha_level = alpha_level,
                                probs = probs)
    }, base_coef_se, resampled_coef_lists)

    ## controversial: in many cases there's just one list, so just send it
    if (length(ci_results) == 1) {
        return(ci_results[[1]])
    }

    ci_results
}

#' Runs the bootstrap estimation method for a single set of coefs (not a list)
#'
#' @inheritParams bootstrap_ci
#' @return Returns a matrix result
#' @keywords internal
bootstrap_individual_ci <- function(base_matrix = NULL,
                                    resampled_coef_list = NULL,
                                    orig_df = NULL,
                                    alpha_level = 0.05,
                                    probs = NULL){
    if (is.null(probs)) {
        if (alpha_level < 0 | alpha_level > 0.5) {
            stop("Can't calculate a two-sided CI with this alpha value, ",
                 "must be in (0, 0.5)")
        }
        probs <- sort(c(alpha_level / 2, 1 - alpha_level / 2))
    }
    if (max(probs) > 1 || min(probs) < 0) {
        stop("Probabilities should be in (0,1)")
    }

    base_row_names <- rownames(base_matrix)
    all_row_names <- lapply(resampled_coef_list, rownames)

    name_match <- unlist(lapply(all_row_names, function(x){
        mean(x == base_row_names) == 1
    }))

    if (mean(name_match) != 1) {
        stop("Naming mismatch from base to list of coefs")
    }

    resampled_ests_vecs <- lapply(1:nrow(base_matrix), function(j){
        unlist(lapply(resampled_coef_list, function(x){
            x[j, 1]
        }))
    })
    resampled_ses_vecs <- lapply(1:nrow(base_matrix), function(j){
        unlist(lapply(resampled_coef_list, function(x){
            x[j, 2]
        }))
    })

    conf_list <- Map(function(base_est,
                              base_se,
                              resampled_ests,
                              resampled_ses){
        ci_variable(base_est,
                    base_se,
                    resampled_ests,
                    resampled_ses,
                    probs = probs)
    },
    base_matrix[, 1],
    base_matrix[, 2],
    resampled_ests_vecs,
    resampled_ses_vecs)

    conf_ind <- do.call(rbind, conf_list)
    rownames(conf_ind) <- rownames(base_matrix) ## only needed for nrow == 1...
    colnames(conf_ind) <- c(
        paste0("boot ",
               names(quantile(1, probs))),
        "boot p_value"
    )

    use_df <- ifelse(is.null(orig_df), Inf, orig_df)
    t_base_vals <- qt(probs, df = use_df)

    if (nrow(base_matrix) > 1) {
        base_mat <- cbind(
            sapply(t_base_vals, function(x){
                base_matrix[, 1] + x * base_matrix[, 2]
            }),
            2 * pt(-abs(base_matrix[, 1] / base_matrix[, 2]),
                   df = use_df)
        )
    } else {
        base_mat <- matrix(c(
            sapply(t_base_vals, function(x){
                base_matrix[, 1] + x * base_matrix[, 2]
            }),
            2 * pt(-abs(base_matrix[, 1] / base_matrix[, 2]),
                   df = use_df)
        ), nrow = 1)
    }

    colnames(base_mat) <- c(
        paste0("base ",
               names(quantile(1, probs))),
        "base p_value"
    )

    if (nrow(base_matrix) > 1) {
        ret_matrix <- cbind(
            "estimate" = base_matrix[, 1],
            round(conf_ind, 4),
            round(base_mat[, c(3, 1, 2)], 4),
            "boot/base width" =
                (conf_ind[, 2] - conf_ind[, 1]) /
                (base_mat[, 2] - base_mat[, 1])
        )
    } else {
        ret_matrix <- matrix(c(
            base_matrix[, 1],
            round(conf_ind, 4),
            round(base_mat[, c(3, 1, 2)], 4),
            (conf_ind[, 2] - conf_ind[, 1]) /
            (base_mat[, 2] - base_mat[, 1])
        ), nrow = 1)
        colnames(ret_matrix) <- c(
            "estimate",
            "boot 2.5%", "boot 97.5%",
            "boot p_value", "base p_value",
            "base 2.5%", "base 97.5%",
            "boot/base width")
        rownames(ret_matrix) <- rownames(base_matrix)
    }

    ret_matrix
}

#' Gets the confidence interval and p-value for a single variable.
#' @param base_est Base model estimate.
#' @param base_se Base model SE.
#' @param resampled_ests Vector of estimates from resampling.
#' @param resampled_ses Vector of standard errors from resampling.
#' @return Returns a length 3 vector: left and right CI value, and p-value.
#' @keywords internal
ci_variable <- function(base_est,
                        base_se,
                        resampled_ests,
                        resampled_ses,
                        probs){
    ## thanks Tim!
    t_boot <- quantile((resampled_ests - base_est) / resampled_ses,
                       1 - probs,
                       type = 6)
    ## can't have the CI not containing the estimate...!
    ## shouldn't happen too often...
    t_boot[1] <- ifelse(t_boot[1] < 0, 0, t_boot[1])
    t_boot[2] <- ifelse(t_boot[2] > 0, 0, t_boot[2])

    result <- base_est - t_boot * base_se

    ## p val:
    p_t <- base_est / base_se

    p_val_num_left <- sum((resampled_ests - base_est) /
                          resampled_ses <=  p_t) + 1
    p_val_num_right <- sum((resampled_ests - base_est) /
                           resampled_ses >=  p_t) + 1

    p_val_num <- 2 * min(p_val_num_left, p_val_num_right)

    p_val <- p_val_num / (length(resampled_ests) + 1)

    c(result, p_val)
}


#' Combines output from multiple bootstrap_model calls
#'
#' If you run glmmboot on e.g. a grid of computers,
#' set \code{return_coefs_instead = TRUE} for each.
#' Then enter them all here. Either just list them out,
#' or put them into one list and enter them.
#' @param ... List of outputs to be combined, or just a bunch of output entries
#'   as separate unnamed arguments.
#' @param return_combined_list Logical, default \code{FALSE}.
#'   \code{TRUE} if you want the combined list of lists,
#'   \code{FALSE} for just the output from bootstrap_ci applied to it.
#' @return Returns the same output as \code{bootstrap_ci} by default,
#'   or the combined list (as if you had just run bootstrap_model once with
#'   all resamples) if \code{return_combined_list = TRUE}
#' @examples
#' \donttest{
#'   data(test_data)
#'   library(glmmTMB)
#'   ## where subj is some RE
#'   test_model <- glmmTMB(y ~ x_var1 + (1 | subj),
#'                         data = test_data,
#'                         family = binomial)
#'   output_list1 <- bootstrap_model(
#'       test_model, base_data = test_data, 99, return_coefs_instead = TRUE)
#'   output_list2 <- bootstrap_model(
#'       test_model, base_data = test_data, 100, return_coefs_instead = TRUE)
#'   output_list3 <- bootstrap_model(
#'       test_model, base_data = test_data, 100, return_coefs_instead = TRUE)
#'   combine_resampled_lists(output_list1, output_list2, output_list3)
#'
#'   num_blocks = 10
#'   num_total_resamples = 299
#'   reg_list <- list()
#'   for(i in 1:num_blocks){
#'       if(i < num_blocks){
#'           block_resamples = floor((num_total_resamples + 1)/num_blocks)
#'       } else {
#'           block_resamples = floor((num_total_resamples + 1)/num_blocks - 1)
#'       }
#'       reg_list[[i]] = bootstrap_model(test_model,
#'                                       base_data = test_data,
#'                                       resamples = block_resamples,
#'                                       return_coefs_instead = TRUE,
#'                                       num_cores = 1) ## increase for parallel
#'   }
#'   boot_ci1 <- combine_resampled_lists(reg_list)
#'   full_list <- combine_resampled_lists(reg_list, return_combined_list = TRUE)
#'   boot_ci2 <- bootstrap_ci(full_list$base_coef_se,
#'                            full_list$resampled_coef_se)
#'   identical(boot_ci1, boot_ci2)
#' }
#'
#' @export
combine_resampled_lists <- function(...,
                                    return_combined_list = FALSE){
    input_lists <- list(...)

    if (length(input_lists) == 1) {
        input_lists <- input_lists[[1]]
    }

    reg_base_coef <- input_lists[[1]]$base_coef_se

    ## now a list of lists of the resampled:
    listlist_resampled <- lapply(input_lists, "[[", 2)
    ## combine to one list
    reg_resampled <- do.call(c, listlist_resampled)

    if (return_combined_list) {
        return(list(base_coef_se = reg_base_coef,
                    resampled_coef_se = reg_resampled))
    }

    bootstrap_ci(base_coef_se = reg_base_coef,
                 resampled_coef_se = reg_resampled)
}
